dll/win32/shlwapi/propbag.cpp

```diff
@@ -0,0 +1,318 @@
+#include <shlwapi.h>
+#include <shlwapi_undoc.h>
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell);
+
+// Dynamic structure array (DSA); See COMCTL32
+struct _DSA
+{
+    INT  nItemCount;
+    LPVOID pData;
+    INT  nMaxCount;
+    INT  nItemSize;
+    INT  nGrow;
+};
+
+// Property Bag Entry
+typedef struct tagPBAGENTRY
+{
+    LPWSTR pszPropName;
+    VARIANT vari;
+} PBAGENTRY, *PPBAGENTRY;
+
+class CBasePropertyBag : public IPropertyBag, public IPropertyBag2
+{
+protected:
+    LONG m_cRefs;   // reference count
+    DWORD m_dwMode; // STGM_* flags
+
+public:
+    CBasePropertyBag(DWORD dwMode)
+        : m_cRefs(1)
+        , m_dwMode(dwMode)
+    {
+    }
+    virtual ~CBasePropertyBag()
+    {
+    }
+
+    // IUnknown interface
+    STDMETHODIMP QueryInterface(REFIID riid, void **ppvObject) override
+    {
+        if (::IsEqualGUID(riid, IID_IPropertyBag2))
+        {
+            AddRef();
+            *ppvObject = dynamic_cast<IPropertyBag2*>(this);
+            return S_OK;
+        }
+        if (::IsEqualGUID(riid, IID_IUnknown) || ::IsEqualGUID(riid, IID_IPropertyBag))
+        {
+            AddRef();
+            *ppvObject = dynamic_cast<IPropertyBag*>(this);
+            return S_OK;
+        }
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP_(ULONG) AddRef() override
+    {
+        return ::InterlockedIncrement(&m_cRefs);
+    }
+    STDMETHODIMP_(ULONG) Release() override
+    {
+        if (::InterlockedDecrement(&m_cRefs) == 0)
+        {
+            delete this;
+            return 0;
+        }
+        return m_cRefs;
+    }
+
+    // IPropertyBag interface (pure)
+    STDMETHODIMP Read(LPCWSTR pszPropName, VARIANT *pvari, IErrorLog *pErrorLog) override = 0;
+    STDMETHODIMP Write(LPCWSTR pszPropName, VARIANT *pvari) override = 0;
+
+    // IPropertyBag2 interface (stubs)
+    STDMETHODIMP Read(
+        ULONG cProperties,
+        PROPBAG2 *pPropBag,
+        IErrorLog *pErrorLog,
+        VARIANT *pvarValue,
+        HRESULT *phrError) override
+    {
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP Write(
+        ULONG cProperties,
+        PROPBAG2 *pPropBag,
+        VARIANT *pvarValue)
+    {
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP CountProperties(ULONG *pcProperties) override
+    {
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP GetPropertyInfo(
+        ULONG iProperty,
+        ULONG cProperties,
+        PROPBAG2 *pPropBag,
+        ULONG *pcProperties) override
+    {
+        return E_NOTIMPL;
+    }
+    STDMETHODIMP LoadObject(
+        LPCWSTR pstrName,
+        DWORD dwHint,
+        IUnknown *pUnkObject,
+        IErrorLog *pErrorLog) override
+    {
+        return E_NOTIMPL;
+    }
+
+    static void* operator new(size_t size)
+    {
+        return ::LocalAlloc(LMEM_FIXED, size);
+    }
+
+    static void operator delete(void* ptr)
+    {
+        ::LocalFree(ptr);
+    }
+};
+
+class CMemPropertyBag : public CBasePropertyBag
+{
+protected:
+    HDSA m_hDSA; // a DSA of PBAGENTRY's
```

Comment:
Why use a DSA, this is c++ code so you can just use one of the high level collections?

Reply:
Windows uses DSA at here. I think that is optimal.

Reply:
I don't see the reason to add a bunch of extra code unless you want to use this from C code, or you need to stay compatible with the memory layout of ms' code (which should not be needed in the internal implementation of something like this, unless this is shared memory that other processes can read or something like it)

Reply:
Used `CStringW`, `CSimpleMap`, and `CComVariant`.

Reply:
Now we do have a lot of additional dependencies. I felt Katayamas initial approach to be superior.

Reply:
The horror, we have less code and it is a lot simpler now....

Reply:
@HBelusca Any idea?

Reply:
If the internals (and layout) of the underlying class that implements the "on-memory" property bag doesn't need to be known by applications using the SHCreatePropertyBagOnMemory API and the associated helpers, then using the ATL may be better here.